### PR TITLE
Api 3601 dynamo ttl

### PR DESCRIPTION
**Pull request by crolarlibertyva**
- Began the adoption of the Dynamo DB Document Client API as opposed to the current usage of the lower-level Database API.
- Also, using this to include the access token expiration as a field with the intent to use it as the "time to live" attribute for the record in the client credentials table.

### CODE REVIEW

**Review by dbrown-va — CHANGES_REQUESTED** [review]
We want the `exp` from the token itself, rather than the `client_assertion`, which will avoid passing the `assert_info`.

I think we can avoid modifying the client credentials strategy as well.

File: oauth-proxy/dynamo_client.js [hunk 1]
```diff
@@ -4,15 +4,26 @@
...
+    dynamoDb = new AWS.DynamoDB({ endpoint: `http://${local}` });
+    AWS.config.update({
+      region: "us-west-2",
+      endpoint: `http://${local}`,
+    });
+    dynamoDb.dbDocClient = new AWS.DynamoDB.DocumentClient();
+  } else {
+    AWS.config.update(awsConfig);
+    dynamoDb = new AWS.DynamoDB();
+    AWS.config.update(awsConfig);
```
Comment (beaugrantham-va) [inline oauth-proxy/dynamo_client.js:22]: Why is this same update statement run twice?

File: oauth-proxy/dynamo_client.js [hunk 2]
```diff
@@ -4,15 +4,26 @@
...
+    dynamoDb.dbDocClient = new AWS.DynamoDB.DocumentClient();
   }
-  return new DynamoDB();
+
+  return dynamoDb;
```
Comment (beaugrantham-va) [inline oauth-proxy/dynamo_client.js:26]: How does this change affect strategies that read/write which were not updated in this PR?

File: oauth-proxy/dynamo_client.js [hunk 3]
```diff
@@ -145,11 +156,31 @@ function saveToDynamoAccessToken(client, accessToken, key, value, TableName) {
   });
 }
 
+function savePayloadToDynamo(dynamoDb, payload, tableName) {
```
Comment (beaugrantham-va) [inline oauth-proxy/dynamo_client.js:159]: Should this function be async since it returns a promise?

File: oauth-proxy/dynamo_client.js (hunk 1 above)
Comment (crolarlibertyva) [reply oauth-proxy/dynamo_client.js:22]: Without the 2nd update, it didn't retain the config settings on the subsequent client substantiation.

File: oauth-proxy/dynamo_client.js (hunk 2 above)
Comment (crolarlibertyva) [reply oauth-proxy/dynamo_client.js:26]: This leaves the original read/writes intact, just needed to add a reference to the document client for use with the expire_on.

Comment (beaugrantham-va) [reply oauth-proxy/dynamo_client.js:26]: Should new code be using `savePayloadToDynamo` and the old functions deprecated?

Comment (crolarlibertyva) [reply oauth-proxy/dynamo_client.js:26]: Yes, will followup with a refactor ticket to address the comment of the module as well.

**Review by beaugrantham-va — APPROVED** [review]

**Review by dbrown-va — APPROVED** [review]

**Comment by dbrown-va** [comment]
Confirmed update to use the expiration from the token (`decodedToken.exp`) itself rather the client assertion.

**Pull request merged**